### PR TITLE
Oecd updates

### PR DESCRIPTION
**Pull request by lara-veylit**

### CODE REVIEW

File: pymrio/tools/ioparser.py [hunk 1]
```diff
@@ -2082,6 +2157,21 @@ def parse_oecd(path, year=None):
     F_unit = pd.DataFrame(
         index=F_factor_input.index, data=mon_unit, columns=IDX_NAMES["unit"]
     )
+    # factor inputs less subsidies for Z 
+    tls = F_factor_input.iloc[:-1]
```
Comment (Asteinvik) [inline pymrio/tools/ioparser.py:2161]: These blocks (line 2160-2174) should have a bit more explaining comments. Its a bit hard to follow the slicing and reshuffling

File: pymrio/tools/ioparser.py [hunk 2]
```diff
@@ -2016,8 +2020,66 @@ def parse_oecd(path, year=None):
     Z_index = pd.MultiIndex.from_tuples(tuple(ll) for ll in Z.index.str.split("_"))
     Z_columns = Z_index.copy()
     Z_index.names = IDX_NAMES["Z_row"]
```
Comment (Asteinvik) [inline pymrio/tools/ioparser.py:2022]: To save some logic you can set the names as a parameter in MultiIndex.from_tuples(). Then you can remove a lot of the copying of index/columns etc

File: pymrio/tools/ioparser.py [hunk 3]
```diff
@@ -2032,12 +2094,25 @@ def parse_oecd(path, year=None):
...
+              "Non-Profit Institutions Serving Households",
+              "General Government Final Consumption",
+              "Gross Fixed Capital Formation",
+              "Changes in Inventories and Valuables",
+              "Direct purchases abroad by residents"] 
+
+
+
+    Y.index = Z.index
+    Y.columns = Y.columns.set_levels(categories, level=1)
```
Comment (Asteinvik) [inline pymrio/tools/ioparser.py:2110]: What is the difference between this index and the index created on line 2095?

File: pymrio/tools/ioutil.py [hunk 4]
```diff
@@ -262,7 +262,8 @@ def diagonalize_columns_to_sectors(
          (all letters are index or header)
             A B     A A A B B B
                     x y z x y z
-        A x 3 1     3 0 0 1 0 0
+        A x 3 1     3 0 0 1 
+        0 0
```
Comment (Asteinvik) [inline pymrio/tools/ioutil.py]: what has happened here?

File: pymrio/tools/ioparser.py [hunk 5]
```diff
@@ -2091,16 +2181,87 @@ def parse_oecd(path, year=None):
...
+
+    """
+
+# This function can only handle the CO2 extension data from the 2021 version
+    if version_year == 2021:
+        raw_ext = pd.read_csv(root_path)
+    else:
+        raise ParserError("Environmental extension is only available for 2021")
+    
+    year_ext = raw_ext.loc[raw_ext["Time"] == io_table_year]
```
Comment (Asteinvik) [inline pymrio/tools/ioparser.py:2229]: When working with data of a certain size, it is often smart to try to overwrite the same variable if it isn't reused. Then you don't need to worry about having 3 almost identical copies of the same data in memory at all times.

File: pymrio/tools/ioparser.py [hunk 6]
```diff
@@ -2091,16 +2181,87 @@ def parse_oecd(path, year=None):
...
+                    'Warehousing and support activities for transportation', 'Postal and courier activities', 'Accommodation and food service activities', 
+                    'Publishing, audiovisual and broadcasting activities', 'Telecommunications', 'IT and other information services', 'Financial and insurance activities', 
+                    'Real estate activities', 'Professional, scientific and technical activities', 'Administrative and support services', 
+                    'Public administration and defence; compulsory social security', 'Education', 'Human health and social work activities', 'Arts, entertainment and recreation', 'Other service activities', 
+                    'Activities of households as employers; undifferentiated goods- and services-producing activities of households for own use']
+
+# From long to wide df
+    ext_clean = ext_clean[ext_clean['Industry'].isin(industries)].copy()
+    ext_clean.drop("Time", axis =1, inplace=True)
+    ext_clean.rename(columns={"COU":"region", "Industry":"sector"}, inplace=True)
```
Comment (Asteinvik) [inline pymrio/tools/ioparser.py:2261]: This renaming should be done right after reading from csv. region and sector are names used elsewhere, so they will make this function easier to understand

File: pymrio/tools/ioparser.py [hunk 7]
```diff
@@ -2091,16 +2181,87 @@ def parse_oecd(path, year=None):
...
+                    'Water supply; sewerage, waste management and remediation activities', 'Construction', 
+                    'Wholesale and retail trade; repair of motor vehicles', 'Land transport and transport via pipelines', 'Water transport', 'Air transport',
+                    'Warehousing and support activities for transportation', 'Postal and courier activities', 'Accommodation and food service activities', 
+                    'Publishing, audiovisual and broadcasting activities', 'Telecommunications', 'IT and other information services', 'Financial and insurance activities', 
+                    'Real estate activities', 'Professional, scientific and technical activities', 'Administrative and support services', 
+                    'Public administration and defence; compulsory social security', 'Education', 'Human health and social work activities', 'Arts, entertainment and recreation', 'Other service activities', 
+                    'Activities of households as employers; undifferentiated goods- and services-producing activities of households for own use']
+
+# From long to wide df
+    ext_clean = ext_clean[ext_clean['Industry'].isin(industries)].copy()
```
Comment (Asteinvik) [inline pymrio/tools/ioparser.py:2259]: I don't think the copy call has any effect here since ext_clean is overwritten

File: pymrio/tools/ioparser.py [hunk 8]
```diff
@@ -2091,16 +2181,87 @@ def parse_oecd(path, year=None):
...
+    ext_clean = year_ext[["COU", "Industry", "Time", "Value"]]
+# only want logical regions (not world, oecd and other non logical rows)
+    regions = ["ARG", 'AUS', 'AUT', 'BEL', 'BGR', 'BRA', 'BRN', 'CAN', 'CHE', 'CHL',
+       'CHN', 'COL', 'CRI', 'CYP', 'CZE', 'DEU', 'DNK', 'ESP', 'EST', 'FIN',
+       'FRA', 'GBR', 'GRC', 'HKG', 'HRV', 'HUN', 'IDN', 'IND', 'IRL', 'ISL',
+       'ISR', 'ITA', 'JPN', 'KAZ', 'KHM', 'KOR', 'LAO', 'LTU', 'LUX', 'LVA',
+       'MAR', 'MEX', 'MLT', 'MMR', 'MYS', 'NLD', 'NOR', 'NZL', 'PER', 'PHL',
+       'POL', 'PRT', 'ROU', 'ROW', 'RUS', 'SAU', 'SGP', 'SVK', 'SVN', 'SWE',
+       'THA', 'TUN', 'TUR', 'TWN', 'USA', 'VNM', 'ZAF']
+    ext_clean = ext_clean[ext_clean['COU'].isin(regions)].copy()
```
Comment (Asteinvik) [inline pymrio/tools/ioparser.py:2241]: unnecessary copy

File: pymrio/tools/ioparser.py [hunk 9]
```diff
@@ -2082,6 +2157,21 @@ def parse_oecd(path, year=None):
     F_unit = pd.DataFrame(
         index=F_factor_input.index, data=mon_unit, columns=IDX_NAMES["unit"]
     )
+    # factor inputs less subsidies for Z 
+    tls = F_factor_input.iloc[:-1]
+    sums = tls.sum(axis = 0)
+    F_factor_input[:1] = sums
+    F_factor_input.rename(index={'AUS_TAXSUB': 'tls'}, inplace=True)
+    F_Z = F_factor_input.iloc[[0, -1]]
```
Comment (Asteinvik) [inline pymrio/tools/ioparser.py:2165]: Why this renaming?

**Comment by Asteinvik** [comment]
By splitting this into multiple functions some of the "dancing" index-renaming can be handled in a more readable way i think

File: pymrio/tools/ioparser.py [hunk 10]
```diff
@@ -2016,8 +2020,66 @@ def parse_oecd(path, year=None):
     Z_index = pd.MultiIndex.from_tuples(tuple(ll) for ll in Z.index.str.split("_"))
     Z_columns = Z_index.copy()
     Z_index.names = IDX_NAMES["Z_row"]
+
+
+    # Make names human readable not codes
+    # Hard coded in May 2023 
+    # Do as excel solution
```
Comment (Asteinvik) [inline pymrio/tools/ioparser.py:2027]: Rather than excel, put these in a .tsv (tab separated) to comply with the structure already done by other databases

File: pymrio/tools/ioparser.py [hunk 11]
```diff
@@ -1882,7 +1882,7 @@ def __get_WIOD_SEA_extension(root_path, year, version, data_sheet="DATA"):
 def parse_oecd(path, year=None):
     """Parse the OECD ICIO tables
 
-    This function works for both, the 2016 and 2018 release.
+    This function works for the 2016, 2018, ad 2021 releases.
```
[the remark this answers is not in the document]
Comment (lara-veylit) [reply pymrio/tools/ioparser.py]: fixed

File: pymrio/tools/ioparser.py (hunk 1 above)
Comment (lara-veylit) [reply pymrio/tools/ioparser.py:2161]: added some explanations here. Basically this block 1) sums all values per country to collapse into one row 2) rename to follow conventions of functions for other databases

File: pymrio/tools/ioparser.py (hunk 3 above)
Comment (lara-veylit) [reply pymrio/tools/ioparser.py:2110]: This needs to be here. It is renaming the columns from codes e.g. "T106" to human readable names i.e. "Gross fixed Capitol Formation"

File: pymrio/tools/ioutil.py (hunk 4 above)
Comment (lara-veylit) [reply pymrio/tools/ioutil.py]: Fixed.

File: pymrio/tools/ioparser.py (hunk 7 above)
Comment (lara-veylit) [reply pymrio/tools/ioparser.py:2259]: deleted the .copy()

File: pymrio/tools/ioparser.py (hunk 8 above)
Comment (lara-veylit) [reply pymrio/tools/ioparser.py:2241]: deleted .copy()

File: pymrio/tools/ioparser.py (hunk 10 above)
Comment (lara-veylit) [reply pymrio/tools/ioparser.py:2027]: To be handled in 2 weeks: 

1) read from tsv instead of hard coded
2) split up functions

File: pymrio/tools/ioparser.py (hunk 9 above)
Comment (lara-veylit) [reply pymrio/tools/ioparser.py:2165]: to be consistent with the other dbs naming conventions

File: pymrio/tools/ioparser.py (hunk 6 above)
Comment (lara-veylit) [reply pymrio/tools/ioparser.py:2261]: good shout. Fixed
